client/rest-high-level/src/main/java/org/elasticsearch/client/security/support/expressiondsl/parser/RoleMapperExpressionParser.java

```diff
@@ -0,0 +1,198 @@
+
+    /**
+     * @param name The name of the expression tree within its containing object.
+     * Used to provide descriptive error messages.
+     * @param parser A parser over the XContent (typically JSON) DSL
+     * representation of the expression
+     */
+    public RoleMapperExpression parse(final String name, final XContentParser parser) throws IOException {
+        return parseRulesObject(name, parser, false);
+    }
+
+    private RoleMapperExpression parseRulesObject(final String objectName, final XContentParser parser, boolean allowExcept)
+            throws IOException {
+        // find the start of the DSL object
+        final XContentParser.Token token;
+        if (parser.currentToken() == null) {
+            token = parser.nextToken();
+        } else {
+            token = parser.currentToken();
+        }
+        if (token != XContentParser.Token.START_OBJECT) {
+            throw new ElasticsearchParseException("failed to parse rules expression. expected [{}] to be an object but found [{}] instead",
+                    objectName, token);
+        }
+
+        final String fieldName = fieldName(objectName, parser);
+        final RoleMapperExpression expr = parseExpression(parser, fieldName, allowExcept, objectName);
+        if (parser.nextToken() != XContentParser.Token.END_OBJECT) {
+            throw new ElasticsearchParseException("failed to parse rules expression. object [{}] contains multiple fields", objectName);
+        }
+        return expr;
+    }
+
+    private RoleMapperExpression parseExpression(XContentParser parser, String field, boolean allowExcept, String objectName)
+            throws IOException {
+
+        if (CompositeType.ANY.getParseField().match(field, parser.getDeprecationHandler())) {
+            CompositeRoleMapperExpression.Builder builder = CompositeRoleMapperExpression.builder(CompositeType.ANY);
+            parseExpressionArray(CompositeType.ANY.getParseField(), parser, false).forEach(builder::addExpression);
+            return builder.build();
+        } else if (CompositeType.ALL.getParseField().match(field, parser.getDeprecationHandler())) {
+            CompositeRoleMapperExpression.Builder builder = CompositeRoleMapperExpression.builder(CompositeType.ALL);
+            parseExpressionArray(CompositeType.ALL.getParseField(), parser, true).forEach(builder::addExpression);
+            return builder.build();
+        } else if (FieldRoleMapperExpression.FieldType.FIELD.getParseField().match(field, parser.getDeprecationHandler())) {
+            return parseFieldExpression(parser);
+        } else if (CompositeType.EXCEPT.getParseField().match(field, parser.getDeprecationHandler())) {
+            if (allowExcept) {
+                return parseExceptExpression(parser);
+            } else {
+                throw new ElasticsearchParseException("failed to parse rules expression. field [{}] is not allowed within [{}]", field,
+                        objectName);
```

Comment:
You can parse it right? So why not.

The point at which there's some potential value is in when you construct actual objects (for a `PUT` role mapping), then if you can prevent an `except` object being a child of anything other than an `all`, then that could help stop a user of the client from making a mistake.
_But_, it's just going to generate a runtime client side error instead of a server side error, so the value is minimal.